nsxt/resource_nsxt_manager_cluster.go

```diff
@@ -0,0 +1,446 @@
+	// This is because the nodes are obtained using cluster client, and does not
+	// contain any information about credential
+	nodeIPToCredentialMap := getIPtoCredentialMap(d)
+
+	var nodeList []map[string]interface{}
+	for _, node := range nodes {
+		_, nodeInSchema := nodeIPToCredentialMap[node.IPAddress]
+		elem := make(map[string]interface{})
+		elem["id"] = node.ID
+		elem["ip_address"] = node.IPAddress
+		// If node is not in schema but in cluster node list,
+		// it means the node needs to be removed.
+		// In this case we don't need its credential because
+		// remove api does not require node credential
+		if nodeInSchema {
+			elem["username"] = nodeIPToCredentialMap[node.IPAddress][0]
+			elem["password"] = nodeIPToCredentialMap[node.IPAddress][1]
+		}
+		elem["fqdn"] = node.Fqdn
+		elem["status"] = node.Status
+		nodeList = append(nodeList, elem)
+	}
+	return d.Set("node", nodeList)
+}
+
+func getIPtoCredentialMap(d *schema.ResourceData) map[string][]string {
+	// returns a map, whose key is IP address of a node,
+	// value is a slice of [node_login_username, node_login_password]
+	nodes := d.Get("node").([]interface{})
+	var idToCredentialMap = map[string][]string{}
+	for _, node := range nodes {
+		data := node.(map[string]interface{})
+		ipAddress := data["ip_address"].(string)
+		userName := data["username"].(string)
+		password := data["password"].(string)
+		cred := []string{userName, password}
+		idToCredentialMap[ipAddress] = cred
+	}
+	return idToCredentialMap
+}
+
+func resourceNsxtManagerClusterCreate(d *schema.ResourceData, m interface{}) error {
+	// Call Joincluster function on nodes that are not in the cluster
+	nodes := getClusterNodesFromSchema(d)
+	if len(nodes) == 0 {
+		return fmt.Errorf("Please provide at least one node ip")
+	} else if len(nodes) == 1 {
+		log.Printf("[INFO] Only one node ip provided, skip joining nodes process")
+		return resourceNsxtManagerClusterRead(d, m)
+	}
+	clusterID, certSha256Thumbprint, hostIP, err := getClusterInfoFromHostNode(d, m)
+	if err != nil {
+		return err
+	}
+
+	for _, guestNode := range nodes {
+		err := joinNodeToCluster(clusterID, certSha256Thumbprint, guestNode, hostIP, d, m)
+		if err != nil {
+			return fmt.Errorf("Failed to join node %s: %s", guestNode.ID, err)
+		}
+	}
+	d.SetId(clusterID)
+	return resourceNsxtManagerClusterRead(d, m)
+}
+
+func getClusterInfoFromHostNode(d *schema.ResourceData, m interface{}) (string, string, string, error) {
+	// function return values are:
+	// clusterID, certSha256Thumbprint, hostIP, error
+	connector := getPolicyConnector(m)
+	client := nsx.NewClusterClient(connector)
+	c := m.(nsxtClients)
+	min := c.CommonConfig.MinRetryInterval
+	max := c.CommonConfig.MaxRetryInterval
+	maxRetries := c.CommonConfig.MaxRetries
+	hostIP := ""
+	for i := 0; i < maxRetries; i++ {
+		clusterConfig, err := client.Get()
+		if err != nil {
+			return "", "", "", fmt.Errorf("Failed to get cluster info from host node: %s", err)
```

Comment:
This error message would not be detailed enough, only error type would be printed out. Please refer to `handleReadError` for reference on printing error details. Same applies to other error messages below.